libgit/autogit_manager.go

```diff
@@ -191,23 +217,54 @@ func (am *AutogitManager) LocalChange(
 }
 
 func (am *AutogitManager) getNodesToInvalidate(
-	affectedNodeIDs []libkbfs.NodeID) (nodes []libkbfs.Node) {
+	affectedNodeIDs []libkbfs.NodeID) (
+	nodes []libkbfs.Node, repoNodeIDs []libkbfs.NodeID) {
 	am.registryLock.RLock()
 	defer am.registryLock.RUnlock()
 	for _, nodeID := range affectedNodeIDs {
 		node, ok := am.repoNodesForWatchedIDs[nodeID]
 		if ok {
 			nodes = append(nodes, node)
+			repoNodeIDs = append(repoNodeIDs, nodeID)
+		}
+	}
+	return nodes, repoNodeIDs
+}
+
+func (am *AutogitManager) clearInvalidatedBrowsers(
+	repoNodeIDs []libkbfs.NodeID) {
+	am.browserLock.Lock()
+	defer am.browserLock.Unlock()
+
+	keys := am.browserCache.Keys()
+	for _, k := range keys {
+		// Clear all cached browsers associated with
+		tmp, ok := am.browserCache.Get(k)
+		if !ok {
+			continue
+		}
+		v, ok := tmp.(browserCacheValue)
+		if !ok {
+			continue
+		}
+		rootNodeID := v.repoFS.RootNode().GetID()
+		for _, nodeID := range repoNodeIDs {
+			if rootNodeID == nodeID {
```

Comment:
Is the idea that rootNode will only be removed when all its children are forgotten already?

Reply:
Not sure which idea you're talking about, or which removal you mean.

`nodeID` represents a block ID that we were notified about in a metadata update.  It doesn't necessarily have to correspond to any `Node` currently in the `NodeCache`.

`rootNodeID` corresponds to a node that is definitely still in the node cache, because `repoFS` holds a reference to it, and also `AutogitManager.repoNodesForWatchedIDs` has a reference to it.

This code is just making sure we wipe all the browser associated with the repo from our browser cache, since the corresponding git repo has been updated.

Reply:
Ah; sorry for the vague question. I was referring to the part where we remove k from browserCache three lines below this, and that whether it was the best place to remove it here when `rootNodeID === nodeID`. I was (mistakenly) imagining that `repoNodeID` would be way more than normally just 1. So your comment above already answered my question. Thanks!